custom_actions.go

```diff
@@ -115,16 +115,26 @@ func collectParameterDefinitions() (arguments []parameterDefinition) {
 
 func checkCustomActionUsage() {
 	var actionUsageRegex = regexp.MustCompile(`(action )?([a-zA-Z0-9]+)\(`)
-	var matches = actionUsageRegex.FindAllStringSubmatch(contents, -1)
-	if len(matches) == 0 {
-		return
-	}
-	for _, match := range matches {
-		var ref = strings.TrimSpace(match[2])
-		if _, found := customActions[ref]; found {
-			customActions[ref].used = true
+
+	customActions[""] = &customAction{body: contents}
+
+	var actionQueue = []string{""}
+	for len(actionQueue) > 0 {
+		var matches = actionUsageRegex.FindAllStringSubmatch(customActions[actionQueue[0]].body, -1)
+		actionQueue = actionQueue[1:]
+		if len(matches) == 0 {
+			continue
+		}
+		for _, match := range matches {
+			var ref = strings.TrimSpace(match[2])
+			if _, found := customActions[ref]; found && !customActions[ref].used {
+				customActions[ref].used = true
+				actionQueue = append(actionQueue, ref)
+			}
 		}
 	}
+
+	delete(customActions, "")
 }
```

Comment:
It seems simpler to change this function to accept content. 

```go
func checkCustomActionUsage(content string) {
	var actionUsageRegex = regexp.MustCompile(`(action )?([a-zA-Z0-9]+)\(`)
	var matches = actionUsageRegex.FindAllStringSubmatch(content, -1)
	if len(matches) == 0 {
		return
	}
	for _, match := range matches {
		var ref = strings.TrimSpace(match[2])
		if _, found := customActions[ref]; found {
			customActions[ref].used = true
		}
	}
}
```

Then in `parseCustomActions()`:

```go
	checkCustomActionUsage(contents)

	for _, action := range customActions {
		if strings.ContainsAny(action.body, "()") {
			checkCustomActionUsage(action.body)
		}
	}
```

Reply:
That makes a lot more sense, thank you!